source/lac/cuda_sparse_matrix.cu

```diff
@@ -0,0 +1,468 @@
+
+  namespace internal
+  {
+    template <typename Number>
+    __global__ void scale(Number                                         *val,
+                          const Number                                    a,
+                          const typename SparseMatrix<Number>::size_type  N)
+    {
+      const typename SparseMatrix<Number>::size_type idx = threadIdx.x +
+                                                           blockIdx.x * blockDim.x;
+      if (idx<N)
+        val[idx] *= a;
+    }
+
+
+
+    void csrmv(cusparseHandle_t handle, bool transpose, int m, int n, int nnz,
+               const cusparseMatDescr_t descr, const float *A_val_dev,
+               const int *A_row_ptr_dev, const int *A_column_index_dev,
+               const float *x, bool add,  float *y)
+    {
+      float alpha = 1.;
+      float beta = add ? 1. : 0.;
+      cusparseOperation_t cusparse_operation = transpose ?
+                                               CUSPARSE_OPERATION_TRANSPOSE :
+                                               CUSPARSE_OPERATION_NON_TRANSPOSE;
+
+      cusparseStatus_t error_code;
+      // This function performs y = alpha*op(A)*x + beta*y
+      error_code = cusparseScsrmv(handle, cusparse_operation, m, n, nnz,
+                                  &alpha, descr, A_val_dev, A_row_ptr_dev,
+                                  A_column_index_dev, x, &beta, y);
+      AssertCusparse(error_code);
+    }
+
+
+
+    void csrmv(cusparseHandle_t handle, bool transpose, int m, int n, int nnz,
+               const cusparseMatDescr_t descr, const double *A_val_dev,
+               const int *A_row_ptr_dev, const int *A_column_index_dev,
+               const double *x, bool add, double *y)
+    {
+      double alpha = 1.;
+      double beta = add ? 1. : 0.;
+      cusparseOperation_t cusparse_operation = transpose ?
+                                               CUSPARSE_OPERATION_TRANSPOSE :
+                                               CUSPARSE_OPERATION_NON_TRANSPOSE;
+
+      cusparseStatus_t error_code;
+      // This function performs y = alpha*op(A)*x + beta*y
+      error_code = cusparseDcsrmv(handle, cusparse_operation, m, n, nnz,
+                                  &alpha, descr, A_val_dev, A_row_ptr_dev,
+                                  A_column_index_dev, x, &beta, y);
+      AssertCusparse(error_code);
+    }
+
+
+
+    template <typename Number>
+    __global__ void l1_norm(const typename SparseMatrix<Number>::size_type n_rows,
+                            const Number *val_dev, const int *column_index_dev,
+                            const int *row_ptr_dev, Number *sums)
+    {
+      const typename SparseMatrix<Number>::size_type row = threadIdx.x +
+                                                           blockIdx.x * blockDim.x;
+
+      if (row<n_rows)
+        {
+          for (int j=row_ptr_dev[row]; j<row_ptr_dev[row+1] ; ++j)
+            sums[column_index_dev[j]] += abs(val_dev[j]);
```

Comment:
Don't you need an atomic add here because several threads might sum into the same index?

Reply:
To add on the previous comment: I guess the test is a bit narrow because the different threads progress in lock step. In the five point stencil, the same column index never appears for the same position in the loop. But it very well do. For example the matrix
```
[ 1 1 0 0 0 ]
[ 1 1 1 0 0 ]
[ 1 0 1 1 0 ]
[ 1 0 0 1 1 ]
[ 1 0 0 0 1 ]
```
should give your norm calculation trouble. Can you test that case and, in case the result is wrong, add an atomic sum?

Reply:
You are right ofc. Thanks!